control-plane-operator/controllers/hostedcontrolplane/hostedcontrolplane_controller.go

```diff
@@ -2898,9 +2899,44 @@ func (r *HostedControlPlaneReconciler) reconcileClusterVersionOperator(ctx conte
 		}
 	}
 
+	if r.EnableCVOManagementClusterMetricsAccess {
+		sa := manifests.ClusterVersionOperatorServiceAccount(hcp.Namespace)
+		if _, err := createOrUpdate(ctx, r.Client, sa, func() error {
+			return cvo.ReconcileServiceAccount(sa, p.OwnerRef)
+		}); err != nil {
+			return fmt.Errorf("failed to reconcile cluster version operator serviceaccount: %w", err)
+		}
+
+		role := manifests.ClusterVersionOperatorRole(hcp.Namespace)
+		if _, err := createOrUpdate(ctx, r.Client, role, func() error {
+			return cvo.ReconcileRole(role, p.OwnerRef)
+		}); err != nil {
+			return fmt.Errorf("failed to reconcile cluster version operator role: %w", err)
+		}
+
+		rb := manifests.ClusterVersionOperatorRoleBinding(hcp.Namespace)
+		if _, err := createOrUpdate(ctx, r.Client, rb, func() error {
+			return cvo.ReconcileRoleBinding(rb, role, p.OwnerRef, hcp.Namespace)
+		}); err != nil {
+			return fmt.Errorf("failed to reconcile cluster version operator rolebinding: %w", err)
+		}
+	} else {
+		if _, err := util.DeleteIfNeeded(ctx, r.Client, manifests.ClusterVersionOperatorServiceAccount(hcp.Namespace)); err != nil {
```

Comment:
Timing on these deletions is a bit fiddly.  You currently have them before the CVO's `ReconcileDeployment`, which means this might remove the service accout out from underneath the outgoing CVO.  Shifting the deletions after the `ReconcileDeployment` would help, but you would still not have insight to the pod-level handoff.  If there's a way to check "is the Deployment level on the target `spec`  I want?", that would be a safe place to put these deletions.  But a bit of a race in an unlikely-to-actually-happen `EnableCVOManagementClusterMetricsAccess`-changed situation is probably not a terrible risk either.

Reply:
I didn't think of that. Thank you for the explanation.

Let's make sure we know whether this logic is needed.

Another solution could be to leave the resources created and only assign the `ServiceAccount` and set the `AutomountServiceAccountToken` when `EnableCVOManagementClusterMetricsAccess == true`. Originally, I wanted to do something similar. Dynamically create the `rules` inside the CVO's `Role`. But, currently,  the defaulting logic doesn't allow to empty the `rules` if something in them already exists, and fiddling with codebase-wide defaulting may be risky.

Reply:
Resolving this discussion per https://github.com/openshift/hypershift/pull/2807#discussion_r1415643634.

I am reverting the deletion changes.

If we do decide, in the future, to support disabling it, we can always come back.